{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dae98805-ed6e-46b1-a10c-d2886f21323d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv\n",
    "import requests\n",
    "load_dotenv()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6ed93081-b28b-4652-aaa3-54e3b5c486fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "import pandas as pd\n",
    "from sqlalchemy import create_engine,text as sql_text\n",
    "\n",
    "\n",
    "db_username = 'myuser'\n",
    "db_password = 'root'\n",
    "db_host = 'localhost'\n",
    "db_port = '5432'\n",
    "db_name = 'medwell_db'\n",
    "\n",
    "connection_string = f'postgresql+psycopg2://{db_username}:{db_password}@{db_host}:{db_port}/{db_name}'\n",
    "engine = create_engine(connection_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c9ac0ef2-e448-4a5b-8ef9-ebcd58fb0935",
   "metadata": {},
   "outputs": [],
   "source": [
    "con=engine.connect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f22a7357-3563-4d18-8591-ca073279969f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "33396d73-6436-4b57-8ff9-f54c153e3638",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.read_sql_query(sql=sql_text('''\n",
    "    select date_of_report,report_type,report_data ,summary from patient_report r join patient_reportdetail d on r.id=d.report_id \n",
    "    order by r.submitted_at desc\n",
    "    limit 5;\n",
    "    '''),con= conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6bf31090-883f-476a-8c3c-8c77719ab915",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date_of_report</th>\n",
       "      <th>report_type</th>\n",
       "      <th>report_data</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3/8/24</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>{'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...</td>\n",
       "      <td>The report shows abnormal values in RBC morpho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3/8/24</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>{'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...</td>\n",
       "      <td>Abnormal values included hemoglobin (12.1 g%),...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13/8/24</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>{'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...</td>\n",
       "      <td>Abnormal values include Hemoglobin, RBC Count,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/09/22</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>{'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...</td>\n",
       "      <td>Hemoglobin, RBC Count, PCV, Neutrophils, and L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  date_of_report   report_type  \\\n",
       "0         3/8/24  blood_report   \n",
       "1         3/8/24  blood_report   \n",
       "2        13/8/24  blood_report   \n",
       "3       11/09/22  blood_report   \n",
       "\n",
       "                                         report_data  \\\n",
       "0  {'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...   \n",
       "1  {'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...   \n",
       "2  {'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...   \n",
       "3  {'pcv': {'max': 50, 'min': 36, 'unit': '%', 'v...   \n",
       "\n",
       "                                             summary  \n",
       "0  The report shows abnormal values in RBC morpho...  \n",
       "1  Abnormal values included hemoglobin (12.1 g%),...  \n",
       "2  Abnormal values include Hemoglobin, RBC Count,...  \n",
       "3  Hemoglobin, RBC Count, PCV, Neutrophils, and L...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "27424619-3be9-4e12-b57b-bf08997a4d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_report_text(report_data):\n",
    "    report_text = \"\"\n",
    "    for test, details in report_data.items():\n",
    "        min_val = details['min']\n",
    "        max_val = details['max']\n",
    "        unit = details['unit']\n",
    "        value = details['value']\n",
    "        if value==-1:continue\n",
    "        \n",
    "        value_text = f\"{value} {unit}\" if value != -1 else \"Not available\"\n",
    "        \n",
    "        report_text += f\"{test.capitalize()}: Normal range {min_val}-{max_val} {unit}, Current value: {value_text}\\n\"\n",
    "    \n",
    "    return report_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8fcc048f-c44f-448b-9b8b-babfded10cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"report_data\"]=df.report_data.apply(generate_report_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75254a8d-a0a5-45be-8107-d8a162d2ebc6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date_of_report</th>\n",
       "      <th>report_type</th>\n",
       "      <th>report_data</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3/8/24</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>Pcv: Normal range 36-50 %, Current value: 38 %...</td>\n",
       "      <td>The report shows abnormal values in RBC morpho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3/8/24</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>Pcv: Normal range 36-50 %, Current value: 38 %...</td>\n",
       "      <td>Abnormal values included hemoglobin (12.1 g%),...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13/8/24</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>Pcv: Normal range 36-50 %, Current value: 31 %...</td>\n",
       "      <td>Abnormal values include Hemoglobin, RBC Count,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/09/22</td>\n",
       "      <td>blood_report</td>\n",
       "      <td>Pcv: Normal range 36-50 %, Current value: 32 %...</td>\n",
       "      <td>Hemoglobin, RBC Count, PCV, Neutrophils, and L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  date_of_report   report_type  \\\n",
       "0         3/8/24  blood_report   \n",
       "1         3/8/24  blood_report   \n",
       "2        13/8/24  blood_report   \n",
       "3       11/09/22  blood_report   \n",
       "\n",
       "                                         report_data  \\\n",
       "0  Pcv: Normal range 36-50 %, Current value: 38 %...   \n",
       "1  Pcv: Normal range 36-50 %, Current value: 38 %...   \n",
       "2  Pcv: Normal range 36-50 %, Current value: 31 %...   \n",
       "3  Pcv: Normal range 36-50 %, Current value: 32 %...   \n",
       "\n",
       "                                             summary  \n",
       "0  The report shows abnormal values in RBC morpho...  \n",
       "1  Abnormal values included hemoglobin (12.1 g%),...  \n",
       "2  Abnormal values include Hemoglobin, RBC Count,...  \n",
       "3  Hemoglobin, RBC Count, PCV, Neutrophils, and L...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1a64098c-99f9-4f81-8a4e-98a15a51c6eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_groq import ChatGroq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "83d97325-96a3-4efe-a565-69b3ec57a6c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm=ChatGroq(api_key=os.environ['GROQ'],model=\"llama3-70b-8192\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "149be72a-8049-4755-b13f-9687864550d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AIMessage(content='Doctor Strange is a fictional superhero appearing in American comic books published by Marvel Comics. He is a sorcerer who uses magical powers to protect the Earth and other dimensions from magical threats.\\n\\nDoctor Strange\\'s real name is Stephen Vincent Strange. He was once a brilliant and arrogant neurosurgeon who lost the use of his hands in a car accident. Desperate to regain the use of his hands, Strange searched for a cure and eventually found himself in the Himalayas, where he encountered the Ancient One, a powerful sorcerer who became his mentor.\\n\\nThe Ancient One taught Strange the mystic arts, and he quickly became a powerful sorcerer in his own right. Using his newfound abilities, Strange created a magical sanctum in New York City and set out to defend the Earth against magical threats.\\n\\nDoctor Strange\\'s powers come from his ability to cast spells, summon magical entities, and wield magical artifacts. He is also a skilled martial artist and strategist. He is often aided by his loyal servant, Wong, and his magical artifact, the Eye of Agamotto.\\n\\nDoctor Strange has been a member of several superhero teams, including the Avengers and the Defenders. He has also had several solo series and has appeared in various Marvel movies, including the Marvel Cinematic Universe (MCU) film \"Doctor Strange\" (2016), played by Benedict Cumberbatch.\\n\\nSome of Doctor Strange\\'s most notable abilities and artifacts include:\\n\\n* The Eye of Agamotto: A powerful magical artifact that allows Strange to see into other dimensions and manipulate time.\\n* The Cloak of Levitation: A magical cloak that allows Strange to fly and teleport.\\n* The Orb of Agamotto: A magical orb that allows Strange to see into other dimensions and predict the future.\\n* The Book of the Vishanti: A powerful magical tome that contains ancient spells and incantations.\\n* The Spell of the Vishanti: A powerful spell that allows Strange to summon the power of the Vishanti, a group of ancient magical beings.\\n\\nOverall, Doctor Strange is a powerful and complex character who has been a part of the Marvel Universe for over 50 years.', response_metadata={'token_usage': {'completion_tokens': 428, 'prompt_tokens': 14, 'total_tokens': 442, 'completion_time': 1.222857143, 'prompt_time': 0.003378949, 'queue_time': 0.011204041000000001, 'total_time': 1.226236092}, 'model_name': 'llama3-70b-8192', 'system_fingerprint': 'fp_87cbfbbc4d', 'finish_reason': 'stop', 'logprobs': None}, id='run-07e758fb-8994-4e24-96b9-bebca9fe3c78-0', usage_metadata={'input_tokens': 14, 'output_tokens': 428, 'total_tokens': 442})"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "llm.invoke(\"who is doctor strange\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "299d8970-e4c6-4b9c-b326-3fccc898c579",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "718b9b69-993e-463b-873a-a4855c44a634",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandasai import SmartDataframe,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9bb16896-bdd0-487d-a96f-5b735917aee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "agent = SmartDataframe(df,config = {'llm':llm})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "41da639f-06d0-465a-a3b9-fa32d199cdbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d02f9b1a-0794-4c88-b8bc-497c7be01a92",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "738319df-fd06-4a75-81e9-0cb228f24762",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 230: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\prompt_generation.py\", line 37, in execute\n",
      "    self.logger.log(f\"Using prompt: {prompt}\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Using prompt: <dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\n\\n\\nUpdate this initial code:\\n```python\\n# TODO: import the required dependencies\\nimport pandas as pd\\n\\n# Write code here\\n\\n# Declare result var: \\ntype (must be \"string\"), value must be string. Example: { \"type\": \"string\", \"value\": f\"The highest salary is {highest_salary}.\" }\\n\\n```\\n\\n\\n\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nVariable `dfs: list[pd.DataFrame]` is already declared.\\n\\nAt the end, declare \"result\" variable as a dictionary of type and value.\\n\\nIf you are asked to plot a chart, use \"matplotlib\" for charts, save as png.\\n\\n\\nGenerate python code and return full updated code:'\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 863: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 36, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Prompt used:\\n            \\n\\n### PREVIOUS CONVERSATION\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n<dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\n\\n\\nUpdate this initial code:\\n```python\\n# TODO: import the required dependencies\\nimport pandas as pd\\n\\n# Write code here\\n\\n# Declare result var: \\ntype (must be \"string\"), value must be string. Example: { \"type\": \"string\", \"value\": f\"The highest salary is {highest_salary}.\" }\\n\\n```\\n\\n\\n\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nVariable `dfs: list[pd.DataFrame]` is already declared.\\n\\nAt the end, declare \"result\" variable as a dictionary of type and value.\\n\\nIf you are asked to plot a chart, use \"matplotlib\" for charts, save as png.\\n\\n\\nGenerate python code and return full updated code:\\n            '\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 281: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 41, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Code generated:\\n            ```\\n            import pandas as pd\\n\\n# Assuming dfs is already a list of DataFrames\\n\\n# Extract RBC count values from each report\\nrbc_counts = [float(df.iloc[0][\\'report_data\\'].split(\\'Rbc_count:\\')[1].split(\\'million cells/μL\\')[0].strip().split(\\'Current value: \\')[1]) for df in dfs]\\n\\n# Calculate average RBC count\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\n\\n# Declare result var\\nresult = {\"type\": \"string\", \"value\": f\"Your average RBC count is {average_rbc_count:.2f} million cells/μL.\"}\\n            ```\\n            '\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 139: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 95, in execute\n",
      "    code_to_run = self.get_code_to_run(input, code_context)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 153, in get_code_to_run\n",
      "    self._logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"\\nCode running:\\n```\\nrbc_counts = [float(df.iloc[0]['report_data'].split('Rbc_count:')[1].split('million cells/μL')[0].strip().split('Current value: ')[1]) for df in dfs]\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n        ```\"\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"<string>\", line 1, in <listcomp>\n",
      "IndexError: list index out of range\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 230: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_prompt_generation.py\", line 56, in execute\n",
      "    self.logger.log(f\"Using prompt: {prompt}\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Using prompt: <dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\nThe user asked the following question:\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nYou generated this python code:\\nrbc_counts = [float(df.iloc[0][\\'report_data\\'].split(\\'Rbc_count:\\')[1].split(\\'million cells/μL\\')[0].strip().split(\\'Current value: \\')[1]) for df in dfs]\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {\\'type\\': \\'string\\', \\'value\\': f\\'Your average RBC count is {average_rbc_count:.2f} million cells/μL.\\'}\\n\\nIt fails with the following error:\\nTraceback (most recent call last):\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 85, in execute\\n    result = self.execute_code(code_to_run, code_context)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 171, in execute_code\\n    exec(code, environment)\\n  File \"<string>\", line 1, in <module>\\n  File \"<string>\", line 1, in <listcomp>\\nIndexError: list index out of range\\n\\n\\nFix the python code above and return the new python code:'\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"<string>\", line 1, in <listcomp>\n",
      "IndexError: list index out of range\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 863: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 36, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Prompt used:\\n            \\n\\n### PREVIOUS CONVERSATION\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n<dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\nThe user asked the following question:\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nYou generated this python code:\\nrbc_counts = [float(df.iloc[0][\\'report_data\\'].split(\\'Rbc_count:\\')[1].split(\\'million cells/μL\\')[0].strip().split(\\'Current value: \\')[1]) for df in dfs]\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {\\'type\\': \\'string\\', \\'value\\': f\\'Your average RBC count is {average_rbc_count:.2f} million cells/μL.\\'}\\n\\nIt fails with the following error:\\nTraceback (most recent call last):\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 85, in execute\\n    result = self.execute_code(code_to_run, code_context)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 171, in execute_code\\n    exec(code, environment)\\n  File \"<string>\", line 1, in <module>\\n  File \"<string>\", line 1, in <listcomp>\\nIndexError: list index out of range\\n\\n\\nFix the python code above and return the new python code:\\n            '\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"<string>\", line 1, in <listcomp>\n",
      "IndexError: list index out of range\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 522: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 41, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"Code generated:\\n            ```\\n            rbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0]['report_data']\\n    for line in report_data.splitlines():\\n        if 'Rbc_count:' in line:\\n            rbc_count = float(line.split(':')[1].strip().split(' ')[0])\\n            rbc_counts.append(rbc_count)\\n            break\\n\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n            ```\\n            \"\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"<string>\", line 1, in <listcomp>\n",
      "IndexError: list index out of range\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 496: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 95, in execute\n",
      "    code_to_run = self.get_code_to_run(input, code_context)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 153, in get_code_to_run\n",
      "    self._logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"\\nCode running:\\n```\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0]['report_data']\\n    for line in report_data.splitlines():\\n        if 'Rbc_count:' in line:\\n            rbc_count = float(line.split(':')[1].strip().split(' ')[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n        ```\"\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 6, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 230: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_prompt_generation.py\", line 56, in execute\n",
      "    self.logger.log(f\"Using prompt: {prompt}\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Using prompt: <dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\nThe user asked the following question:\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nYou generated this python code:\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0][\\'report_data\\']\\n    for line in report_data.splitlines():\\n        if \\'Rbc_count:\\' in line:\\n            rbc_count = float(line.split(\\':\\')[1].strip().split(\\' \\')[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {\\'type\\': \\'string\\', \\'value\\': f\\'Your average RBC count is {average_rbc_count:.2f} million cells/μL.\\'}\\n\\nIt fails with the following error:\\nTraceback (most recent call last):\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 85, in execute\\n    result = self.execute_code(code_to_run, code_context)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 171, in execute_code\\n    exec(code, environment)\\n  File \"<string>\", line 6, in <module>\\nValueError: could not convert string to float: \\'Normal\\'\\n\\n\\nFix the python code above and return the new python code:'\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 6, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 863: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 36, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Prompt used:\\n            \\n\\n### PREVIOUS CONVERSATION\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n<dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\nThe user asked the following question:\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nYou generated this python code:\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0][\\'report_data\\']\\n    for line in report_data.splitlines():\\n        if \\'Rbc_count:\\' in line:\\n            rbc_count = float(line.split(\\':\\')[1].strip().split(\\' \\')[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {\\'type\\': \\'string\\', \\'value\\': f\\'Your average RBC count is {average_rbc_count:.2f} million cells/μL.\\'}\\n\\nIt fails with the following error:\\nTraceback (most recent call last):\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 85, in execute\\n    result = self.execute_code(code_to_run, code_context)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 171, in execute_code\\n    exec(code, environment)\\n  File \"<string>\", line 6, in <module>\\nValueError: could not convert string to float: \\'Normal\\'\\n\\n\\nFix the python code above and return the new python code:\\n            '\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 6, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 623: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 41, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"Code generated:\\n            ```\\n            rbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0]['report_data']\\n    for line in report_data.splitlines():\\n        if 'Rbc_count:' in line:\\n            parts = line.split(':')\\n            value_part = parts[1].strip()\\n            value_parts = value_part.split(' ')\\n            rbc_count = float(value_parts[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n            ```\\n            \"\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 6, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 599: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 95, in execute\n",
      "    code_to_run = self.get_code_to_run(input, code_context)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 153, in get_code_to_run\n",
      "    self._logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"\\nCode running:\\n```\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0]['report_data']\\n    for line in report_data.splitlines():\\n        if 'Rbc_count:' in line:\\n            parts = line.split(':')\\n            value_part = parts[1].strip()\\n            value_parts = value_part.split(' ')\\n            rbc_count = float(value_parts[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n        ```\"\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 9, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 230: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_prompt_generation.py\", line 56, in execute\n",
      "    self.logger.log(f\"Using prompt: {prompt}\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Using prompt: <dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\nThe user asked the following question:\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nYou generated this python code:\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0][\\'report_data\\']\\n    for line in report_data.splitlines():\\n        if \\'Rbc_count:\\' in line:\\n            parts = line.split(\\':\\')\\n            value_part = parts[1].strip()\\n            value_parts = value_part.split(\\' \\')\\n            rbc_count = float(value_parts[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {\\'type\\': \\'string\\', \\'value\\': f\\'Your average RBC count is {average_rbc_count:.2f} million cells/μL.\\'}\\n\\nIt fails with the following error:\\nTraceback (most recent call last):\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 85, in execute\\n    result = self.execute_code(code_to_run, code_context)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 171, in execute_code\\n    exec(code, environment)\\n  File \"<string>\", line 9, in <module>\\nValueError: could not convert string to float: \\'Normal\\'\\n\\n\\nFix the python code above and return the new python code:'\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 9, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 863: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 36, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: 'Prompt used:\\n            \\n\\n### PREVIOUS CONVERSATION\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n<dataframe>\\ndfs[0]:4x4\\ndate_of_report,report_type,report_data,summary\\r\\n3/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 32 %\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.98 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 5100 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.8 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 15 %\\nNeutrophils: Normal range 40-75 %, Current value: 80 %\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 159000 cells/μL\\n\",\"Hemoglobin, RBC Count, PCV, Neutrophils, and Lymphocytes are abnormal.\"\\r\\n11/09/22,blood_report,\"Pcv: Normal range 36-50 %, Current value: 38 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.16 g/dL\\nCalcium: Normal range 8.5-10.2 mg/dL, Current value: 8.19 mg/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 3.02 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.18 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 4.72 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 9700 thousand cells/μL\\nBlood_urea: Normal range 7-20 mg/dL, Current value: 21 mg/dL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 12.1 g/dL\\nPhosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nLymphocytes: Normal range 20-45 %, Current value: 6 %\\nNeutrophils: Normal range 40-75 %, Current value: 91 %\\nS_uric_acid: Normal range 3.5-7.2 mg/dL, Current value: 4.25 mg/dL\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.98 mg/dL\\nS_phosphorus: Normal range 2.5-4.5 mg/dL, Current value: 3.84 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 219000 cells/μL\\nBilirubin_total: Normal range 0.1-1.2 mg/dL, Current value: 0.57 mg/dL\\nBilirubin_direct: Normal range 0.0-0.3 mg/dL, Current value: 0.19 mg/dL\\nBilirubin_indirect: Normal range 0.1-0.8 mg/dL, Current value: 0.38 mg/dL\\nBlood_urea_nitrogen: Normal range 7-20 mg/dL, Current value: 9.81 mg/dL\\n\",\"Abnormal values include Hemoglobin, RBC Count, PCV, Proteins, Albumin, Globulin, S. Creatinine, Neutrophils, and Lymphocytes.\"\\r\\n13/8/24,blood_report,\"Pcv: Normal range 36-50 %, Current value: 31 %\\nAlbumin: Normal range 3.5-5.0 g/dL, Current value: 3.1 g/dL\\nGlobulin: Normal range 2.0-4.0 g/dL, Current value: 2.97 g/dL\\nProteins: Normal range 6.0-8.0 g/dL, Current value: 6.07 g/dL\\nRbc_count: Normal range 4.2-5.4 million cells/μL, Current value: 3.85 million cells/μL\\nWbc_count: Normal range 4.5-11.0 thousand cells/μL, Current value: 7800 thousand cells/μL\\nHemoglobin: Normal range 12.1-15.5 g/dL, Current value: 10.9 g/dL\\nLymphocytes: Normal range 20-45 %, Current value: 25 %\\nNeutrophils: Normal range 40-75 %, Current value: 67 %\\nS_creatinine: Normal range 0.6-1.2 mg/dL, Current value: 0.99 mg/dL\\nPlatelet_count: Normal range 150000-450000 cells/μL, Current value: 305000 cells/μL\\n\",\"Abnormal values included hemoglobin (12.1 g%), RBC count (4.72 million/cu.mm), PCV (38%), platelet count (219000/cu.mm), bilirubin indirect (0.38 mg/dl), S. Creatinine (0.98 mg/dl), S. Phosphorus (3.84 mg/dl), and calcium (8.19 mg/dl).\"\\r\\n</dataframe>\\n\\n\\nThe user asked the following question:\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n what is the summary of the recent report\\n### ANSWER\\n The report shows abnormal values in RBC morphology, Platelet count, Neutrophils, and Globulin levels ...\\n### QUERY\\n explain my hemoglobin trend\\n### QUERY\\n explain my hemoglobin trend in text\\n### ANSWER\\n Your hemoglobin level is 12.1 g/dL.\\n### QUERY\\n what is my average rbc count\\n### QUERY\\n what is my average rbc count.Give a single value\\n### QUERY\\n what is my average rbc count.Give a single value\\n\\nYou generated this python code:\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0][\\'report_data\\']\\n    for line in report_data.splitlines():\\n        if \\'Rbc_count:\\' in line:\\n            parts = line.split(\\':\\')\\n            value_part = parts[1].strip()\\n            value_parts = value_part.split(\\' \\')\\n            rbc_count = float(value_parts[0])\\n            rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {\\'type\\': \\'string\\', \\'value\\': f\\'Your average RBC count is {average_rbc_count:.2f} million cells/μL.\\'}\\n\\nIt fails with the following error:\\nTraceback (most recent call last):\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 85, in execute\\n    result = self.execute_code(code_to_run, code_context)\\n             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\\n  File \"B:\\\\Major Project\\\\major\\\\Lib\\\\site-packages\\\\pandasai\\\\pipelines\\\\chat\\\\code_execution.py\", line 171, in execute_code\\n    exec(code, environment)\\n  File \"<string>\", line 9, in <module>\\nValueError: could not convert string to float: \\'Normal\\'\\n\\n\\nFix the python code above and return the new python code:\\n            '\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 9, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 679: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_generator.py\", line 41, in execute\n",
      "    logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"Code generated:\\n            ```\\n            rbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0]['report_data']\\n    for line in report_data.splitlines():\\n        if 'Rbc_count:' in line:\\n            parts = line.split(':')\\n            value_part = parts[1].strip()\\n            if 'Normal' not in value_part:\\n                value_parts = value_part.split(' ')\\n                rbc_count = float(value_parts[0])\\n                rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n            ```\\n            \"\n",
      "Arguments: ()\n",
      "--- Logging error ---\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 9, in <module>\n",
      "ValueError: could not convert string to float: 'Normal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\logging\\__init__.py\", line 1113, in emit\n",
      "    stream.write(msg + self.terminator)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\encodings\\cp1252.py\", line 19, in encode\n",
      "    return codecs.charmap_encode(input,self.errors,encoding_table)[0]\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "UnicodeEncodeError: 'charmap' codec can't encode character '\\u03bc' in position 655: character maps to <undefined>\n",
      "Call stack:\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\chouh\\AppData\\Local\\Temp\\ipykernel_11500\\11785461.py\", line 1, in <module>\n",
      "    resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\smart_dataframe\\__init__.py\", line 131, in chat\n",
      "    return self._agent.chat(query, output_type)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\agent\\base.py\", line 274, in chat\n",
      "    return self.pipeline.run(pipeline_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 125, in execute\n",
      "    code_to_run = self._retry_run_code(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 346, in _retry_run_code\n",
      "    return self.on_retry(code, e)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 162, in on_code_retry\n",
      "    return self.code_exec_error_pipeline.run(correction_input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\error_correction_pipeline\\error_correction_pipeline.py\", line 48, in run\n",
      "    return self.pipeline.run(input)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 95, in execute\n",
      "    code_to_run = self.get_code_to_run(input, code_context)\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_cleaning.py\", line 153, in get_code_to_run\n",
      "    self._logger.log(\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\helpers\\logger.py\", line 76, in log\n",
      "    self._logger.info(message)\n",
      "Message: \"\\nCode running:\\n```\\nrbc_counts = []\\nfor df in dfs:\\n    report_data = df.iloc[0]['report_data']\\n    for line in report_data.splitlines():\\n        if 'Rbc_count:' in line:\\n            parts = line.split(':')\\n            value_part = parts[1].strip()\\n            if 'Normal' not in value_part:\\n                value_parts = value_part.split(' ')\\n                rbc_count = float(value_parts[0])\\n                rbc_counts.append(rbc_count)\\n            break\\naverage_rbc_count = sum(rbc_counts) / len(rbc_counts)\\nresult = {'type': 'string', 'value': f'Your average RBC count is {average_rbc_count:.2f} million cells/μL.'}\\n        ```\"\n",
      "Arguments: ()\n",
      "Traceback (most recent call last):\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\generate_chat_pipeline.py\", line 335, in run\n",
      "    ).run(input)\n",
      "      ^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 137, in run\n",
      "    raise e\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\pipeline.py\", line 101, in run\n",
      "    step_output = logic.execute(\n",
      "                  ^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 113, in execute\n",
      "    raise e\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 85, in execute\n",
      "    result = self.execute_code(code_to_run, code_context)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"B:\\Major Project\\major\\Lib\\site-packages\\pandasai\\pipelines\\chat\\code_execution.py\", line 171, in execute_code\n",
      "    exec(code, environment)\n",
      "  File \"<string>\", line 13, in <module>\n",
      "ZeroDivisionError: division by zero\n"
     ]
    }
   ],
   "source": [
    "resp=agent.chat('what is my average rbc count.Give a single value',output_type=\"string\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "1315f766-8062-484a-b4a9-cc725e382849",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Unfortunately, I was not able to answer your question, because of the following error:\\n\\ndivision by zero\\n'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0162e811-6bb7-48bf-b767-f239dc172955",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b647f28c-f707-4b2d-a469-ab1efbdcc0e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "major",
   "language": "python",
   "name": "major"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
